restapi/src/main/java/io/stargate/web/docsapi/service/QueryExecutor.java

```diff
@@ -57,30 +74,109 @@ public QueryExecutor(DataStore dataStore) {
       int pageSize,
       ByteBuffer pagingState,
       ExecutionContext context) {
+    return queryDocs(keyDepth, ImmutableList.of(query), pageSize, pagingState, context);
+  }
+
+  public Flowable<RawDocument> queryDocs(
+      int keyDepth,
+      List<BoundQuery> queries,
+      int pageSize,
+      ByteBuffer pagingState,
+      ExecutionContext context) {
     if (pageSize <= 0) {
       // Note: if page size is not set, C* will ignore paging state, but subsequent pages may be
       // requested by the execute() method, so we require a specific page size for all queries here.
       throw new IllegalArgumentException("Unsupported page size: " + pageSize);
     }
 
-    BuiltSelect select = (BuiltSelect) query.source().query();
-    if (keyDepth < 1 || keyDepth > select.table().primaryKeyColumns().size()) {
-      throw new IllegalArgumentException("Invalid document identity depth: " + keyDepth);
-    }
+    List<Column> idColumns = idColumns(keyDepth, queries);
+    Comparator<DocProperty> comparator = rowComparator(queries);
 
-    List<Column> idColumns = select.table().primaryKeyColumns().subList(0, keyDepth);
+    CombinedPagingState pagingStates = CombinedPagingState.deserialize(queries.size(), pagingState);
 
-    return execute(query, pageSize, pagingState)
-        .flatMap(
-            rs -> Flowable.fromIterable(seeds(query, rs, idColumns, context)),
-            1) // concurrency factor 1
+    PagingStateTracker tracker = new PagingStateTracker(pagingStates);
+
+    return execute(queries, comparator, pageSize, pagingStates, context)
+        .map(p -> toSeed(p, comparator, idColumns))
         .concatWith(Single.just(TERM))
-        .scan(Accumulator::combine)
+        .scan(tracker::combine)
         .filter(Accumulator::isComplete)
         .map(Accumulator::toDoc);
   }
 
+  private Comparator<DocProperty> rowComparator(List<BoundQuery> queries) {
+    List<Column> pathColumns = null;
+    for (BoundQuery query : queries) {
+      BuiltSelect select = (BuiltSelect) query.source().query();
+
+      if (pathColumns == null) {
+        pathColumns = select.table().clusteringKeyColumns();
+      }
+    }
+
+    if (pathColumns == null) {
+      throw new IllegalArgumentException("No query provided");
+    }
+
+    Comparator<DocProperty> comparator = Comparator.comparing(DocProperty::comparableKey);
+    for (Column column : pathColumns) {
+      comparator = comparator.thenComparing(p -> p.keyValue(column));
```

Comment:
why we are including this extra? Isn't enough that the order is per key? Why do we care for the other sub-document props here? Who cares how are the sub-documents of the single document order, that's all on the same partition?

Reply:
We need strong order to be able to identify duplicate rows without loading all of the data into memory. Currently we assume all queries have the same sorting order (according to `pNNN` clustering columns). Then, we can merge the result sets by comparing only the top-level elements.

Reply:
Hmm very good point.. I did not think about the duplication of the rows from the different queries..

So how is this handled? Where are duplicates dismissed? I guess `orderedMerge` will not do that? Is this something we handle in the Accumulator..

Reply:
`orderedMerge` will merge source flows placing equal rows next to each other. `Accumulator.combine(...)` later will take only one of a group of equal rows.